waf_dl.go

```diff
@@ -44,42 +43,53 @@ func dumpWafLibrary() (*os.File, error) {
 	return file, nil
 }
 
-// newWafDl loads the waf shared library and loads all the needed symbols
-func newWafDl() (_ *wafDl, err error) {
+// newWafDl loads the libddwaf shared library along with all the needed symbols.
+// The returned dynamic library handle dl can be non-nil even with a returned
+// error, meaning that the dynamic library handle can be used but some errors
+// happened in the last internal steps following the successful call to
+// dlopen().
+func newWafDl() (dl *wafDl, err error) {
 	file, err := dumpWafLibrary()
 	if err != nil {
 		return nil, err
 	}
-
+	fName := file.Name()
 	defer func() {
-		// TODO(eliott.bouhana): find a way to send to the tracer some non-fatal errors while opening the waf libg
-		file.Close()
-		os.Remove(file.Name())
+		rmErr := os.Remove(fName)
+		if err == nil {
+			err = rmErr
+		} else {
+			// TODO: rely on errors.Join() once go1.20 is our min supported Go version
+			err = fmt.Errorf("%w; along with an error while removing %s: %v", err, fName, rmErr)
```

Comment:
What if `err` is set but there was no error returned by `os.Remove(fName)`? Seems to me like the message will be inaccurate and we could be printing a nil `rmErr`.

Reply:
Good catch, that was a mistake indeed, fixed in 73948c9
A shame we can hardly test those lines, unless relying on interface abstractions to mock them in tests -_-